src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1110 @@
+// dispatch and wouldn't gain us much anyways: the probability of finding a
+// match drops off drastically after the first few buckets.
+//
+// I attempted an implementation on ARM using NEON instructions, but it turns
+// out that most NEON instructions have multi-cycle latency, which in the end
+// outweighs any gains over the generic implementation.
+#[cfg(all(
+    not(stage0),
+    target_feature = "sse2",
+    any(target_arch = "x86", target_arch = "x86_64")
+))]
+#[path = "sse2.rs"]
+mod imp;
+#[cfg(not(all(
+    not(stage0),
+    target_feature = "sse2",
+    any(target_arch = "x86", target_arch = "x86_64")
+)))]
+#[path = "generic.rs"]
+mod imp;
+
+mod bitmask;
+
+use self::bitmask::BitMask;
+use self::imp::Group;
+
+/// Whether memory allocation errors should return an error or abort.
+enum Fallibility {
+    Fallible,
+    Infallible,
+}
+
+impl Fallibility {
+    /// Error to return on capacity overflow.
+    #[inline]
+    fn capacity_overflow(&self) -> CollectionAllocErr {
+        match *self {
+            Fallibility::Fallible => CollectionAllocErr::CapacityOverflow,
+            Fallibility::Infallible => panic!("Hash table capacity overflow"),
+        }
+    }
+
+    /// Error to return on allocation error.
+    #[inline]
+    fn alloc_err(&self, layout: Layout) -> CollectionAllocErr {
+        match *self {
+            Fallibility::Fallible => CollectionAllocErr::AllocErr,
+            Fallibility::Infallible => handle_alloc_error(layout),
+        }
+    }
+}
+
+/// Control byte value for an empty bucket.
+const EMPTY: u8 = 0b11111111;
+
+/// Control byte value for a deleted bucket.
+const DELETED: u8 = 0b10000000;
+
+/// Checks whether a control byte represents a full bucket (top bit is clear).
+#[inline]
+fn is_full(ctrl: u8) -> bool {
+    ctrl & 0x80 == 0
+}
+
+/// Checks whether a control byte represents a special value (top bit is set).
+#[inline]
+fn is_special(ctrl: u8) -> bool {
+    ctrl & 0x80 != 0
+}
+
+/// Checks whether a special control value is EMPTY (just check 1 bit).
+#[inline]
+fn special_is_empty(ctrl: u8) -> bool {
+    debug_assert!(is_special(ctrl));
+    ctrl & 0x01 != 0
+}
+
+/// Primary hash function, used to select the initial bucket to probe from.
+#[inline]
+fn h1(hash: u64) -> usize {
+    hash as usize
+}
+
+/// Secondary hash function, saved in the low 7 bits of the control byte.
+#[inline]
+fn h2(hash: u64) -> u8 {
+    // Grab the top 7 bits of the hash. While the hash is normally a full 64-bit
+    // value, some hash functions (such as FxHash) produce a usize result
+    // instead, which means that the top 32 bits are 0 on 32-bit platforms.
+    let hash_len = usize::min(mem::size_of::<usize>(), mem::size_of::<u64>());
+    let top7 = hash >> (hash_len * 8 - 7);
+    (top7 & 0x7f) as u8
+}
+
+/// Probe sequence based on triangular numbers, which is guaranteed (since our
+/// table size is a power of two) to visit every group of elements exactly once.
```

Comment:
Suppose the probe sequence *T*(*n*) ≡ (*n*² + *n*)/2 (mod 2<sup>*k*</sup>) repeats a group before visiting all 2<sup>*k*</sup> groups: that is, (*a*² + *a*)/2 ≡ (*b*² + *b*) (mod 2<sup>*k*</sup>) for some 0 ≤ *a* < *b* < 2<sup>*k*</sup>. Then (*b* − *a*)(*a* + *b* + 1) = *b*² + *b* − *a*² − *a* is a multiple of 2<sup>*k* + 1</sup>. Since its factors *b* − *a* and *a* + *b* + 1 are of different parity, one of them is itself a multiple of 2<sup>*k* + 1</sup>. But both are positive and too small to be a multiple of 2<sup>*k* + 1</sup>, so that’s a contradiction.